### ISSUE TITLE
error: Package ‘python-2.7.18.6’ in cpython/2.7/default.nix:341 is marked as insecure, refusing to evaluate.

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#210098]
### Describe the bug

I want to update my system and get this error:

> error: Package ‘python-2.7.18.6’ in /nix/store/vb75r2a5cdcxfc3ba4zfc1gwdfnb3ra8-nixos-unstable/nixos-unstable/pkgs/development/interpreters/python/cpython/2.7/default.nix:341 is marked as insecure, refusing to evaluate.
> 
> 
>        Known issues:
>         - Python 2.7 has reached its end of life after 2020-01-01. See https://www.python.org/doc/sunset-python-2/.
> 
>        You can install it anyway by allowing this package, using the
>        following methods:
> 
>        a) To temporarily allow all insecure packages, you can use an environment
>           variable for a single invocation of the nix tools:
> 
>             $ export NIXPKGS_ALLOW_INSECURE=1
> 
>         Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
>         (Flake) command, `--impure` must be passed in order to read this
>         environment variable.
> 
>        b) for `nixos-rebuild` you can add ‘python-2.7.18.6’ to
>           `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
>           like so:
> 
>             {
>               nixpkgs.config.permittedInsecurePackages = [
>                 "python-2.7.18.6"
>               ];
>             }
> 
>        c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
>           ‘python-2.7.18.6’ to `permittedInsecurePackages` in
>           ~/.config/nixpkgs/config.nix, like so:
> 
>             {
>               permittedInsecurePackages = [
>                 "python-2.7.18.6"
>               ];
>             }

Caused by https://github.com/NixOS/nixpkgs/pull/201859

### Steps To Reproduce
1. `nixos-rebuild switch`
2. get error

### Expected behavior

In best case, i would not get errors.

But if i get error messages, they should tell me how to resolve the issue.
This one only talks about allowing insecure packages, but i want to consider removing the causing package from my system.
The thing is, python2 is not a package i install, so it has to be a dependency of some package. The error message should also explain which package i have to remove to fix the error!

### Additional context

Without a solution, i'm blocked to update my system.

It looks like a solution is technically not possible: https://discourse.nixos.org/t/why-does-my-system-config-depend-on-now-broken-package/2654

### Workaround

Build with temporarily allowing the insecure package.

`export NIXPKGS_ALLOW_INSECURE=1`

Then check why your built system depends on python2 in this case.

```
[root@gaming:~]# NIXPKGS_ALLOW_INSECURE=1 nix why-depends /nix/store/lryhdsyzrjb1j2gbvy45jvzg8pmly00s-nixos-system-gaming-22.11.1482.8c54d842d95.drv github:nixos/nixpkgs/nixos-unstable#python2 --impure
/nix/store/wmlcib5fnl6ldjb78mlfcx35kn7kylp3-nixos-system-gaming-22.11.1482.8c54d842d95
└───/nix/store/0s9h935yiwjbbgwj229zfxrx1s26yxk9-system-path
    └───/nix/store/sy77zp1p5hvmdmgjw4b4z66rmpm27swb-davinci-resolve
        └───/nix/store/vcj8psnh3hy9dqng435hiqp0qwmywhgc-davinci-resolve-init
            └───/nix/store/25q5a887s0dr8cv50ccrdflw7vykpk7k-davinci-resolve-fhs
                └───/nix/store/6yk94gpz7iqimfy22ivr7w1j0znkmnz1-davinci-resolve-usr-target
                    └───/nix/store/srrxbcf4bsjqzr2k44jml0rhkwz04h9k-python-2.7.18.6
```

### Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.1, NixOS, 22.11 (Raccoon), 22.11.1137.dac57a4eccf`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - channels(root): `"home-manager-22.11.tar.gz, nixos-22.11, nixos-hardware, nixos-unstable"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by abeforgit** [comment]
I was in the same boat as you, while looking around I found this very handy tool: https://github.com/utdemir/nix-tree
you can pass `--derivation` as an option to look at build-time dependencies. (I recommend searching both with and without the flag) You can search with the `/` key and navigate with arrow keys or hjkl. Also useful is the `w` key which shows you the why-depends of the highlighted package, aka a list of all packages that depend on it.

Important: search for `python-2` and not `python2` ,which I did at first and lost me quite some time.

Also, in my case the culprit was a build dependency of  `nvtop`, which is quite a common tool to have so maybe it's the same for you

**Comment by fabianhjr** [comment]
There is a similar issue focusing on the error itself rather than the package marked as insecure: https://github.com/NixOS/nixpkgs/issues/209804

>  "Package ... is marked as insecure" messages could be more actionable

**Comment by markus1189** [comment]
@abeforgit Thanks for the tip about nix-tree!

I used:

```
nix run github:utdemir/nix-tree -- /nix/var/nix/profiles/system
```

**Comment by ju1m** [comment]
Think also to look for `\<python\>` (or `\<pythonFull\>`) (eg. in `environment.systemPackages`) because since https://github.com/NixOS/nixpkgs/pull/154687 it's an alias for `python-2`.

**Comment by woylie** [comment]
> nix run github:utdemir/nix-tree -- /nix/var/nix/profiles/system

I ran that, got into search mode with `/`, typed `python`, and all I see are `python3*` packages. I don't have any references to Python 2 in my `configuration.nix` either. Not sure what to do here, except for `export NIXPKGS_ALLOW_INSECURE=1` whenever I upgrade my system.

**Comment by pbsds** [comment]
OPs case is fixed in #240549, closing in favor of #209804
